runtime/compiler/runtime/StatisticsThread.cpp

```diff
@@ -0,0 +1,197 @@
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "runtime/StatisticsThread.hpp"
+#include "env/VMJ9.h"
+#include "runtime/CompileService.hpp"
+
+#define SYNC_TIME 100
+#define STATISTICS_LOG_PERIOD 2000
+TR_StatisticsThread::TR_StatisticsThread()
+   : _statisticsThread(NULL), _statisticsThreadMonitor(NULL), _statisticsOSThread(NULL),
+   _statisticsThreadAttachAttempted(false), _statisticsThreadExitFlag(false)
+   {
+   }
+
+TR_StatisticsThread * TR_StatisticsThread::allocate()
+   {
+   TR_StatisticsThread * statisticsThread = new (PERSISTENT_NEW) TR_StatisticsThread();
+   return statisticsThread;
+   }
+
+static int32_t J9THREAD_PROC statisticsThreadProc(void * entryarg)
+   {
+   J9JITConfig * jitConfig = (J9JITConfig *) entryarg;
+   J9JavaVM * vm = jitConfig->javaVM;
+
+   TR_StatisticsThread *statisticsThread = ((TR_JitPrivateConfig*)(jitConfig->privateConfig))->statisticsThread;
+   UDATA samplingPeriod    = std::max(static_cast<UDATA>(TR::Options::_minSamplingPeriod), jitConfig->samplingFrequency);
+
+   J9VMThread *statThread = NULL;
+   PORT_ACCESS_FROM_JAVAVM(vm);
+
+   int rc = vm->internalVMFunctions->internalAttachCurrentThread(vm, &statThread, NULL,
+                                  J9_PRIVATE_FLAGS_DAEMON_THREAD | J9_PRIVATE_FLAGS_NO_OBJECT |
+                                  J9_PRIVATE_FLAGS_SYSTEM_THREAD | J9_PRIVATE_FLAGS_ATTACHED_THREAD,
+                                  statisticsThread->getStatisticsOSThread());
+   if (rc != JNI_OK)
+      return JNI_ERR; // attaching the JITaaS Statistics thread failed
+
+   statisticsThread->getStatisticsThreadMonitor()->enter();
+   statisticsThread->setAttachAttempted(true);
+   statisticsThread->setStatisticsThread(statThread);
+   statisticsThread->getStatisticsThreadMonitor()->notifyAll();
+   statisticsThread->getStatisticsThreadMonitor()->exit();
+
+   j9thread_set_name(j9thread_self(), "JITaaS Server Statistics Thread");
+
+   TR::CompilationInfo *compInfo = TR::CompilationInfo::get(jitConfig);
+
+   TR::PersistentInfo *persistentInfo = compInfo->getPersistentInfo();
+   persistentInfo->setStartTime(j9time_current_time_millis());
+   persistentInfo->setElapsedTime(0);
+
+   uint64_t oldSyncTime = 0;
+   uint64_t crtTime = 0;
+   uint64_t lastStatistics = 0;
+   uint64_t syncDrift = 0;
+   while(!statisticsThread->getStatisticsThreadExitFlag())
+      {
+      while(!statisticsThread->getStatisticsThreadExitFlag() && j9thread_sleep_interruptable((IDATA) samplingPeriod - syncDrift, 0) == 0)
+         {
+         persistentInfo->updateElapsedTime(samplingPeriod);
+         crtTime += samplingPeriod;
+
+         //adjusting the sampling period window for sync drift.
```

Comment:
I do not understand how time correction can be done without reading the time from the OS.

Reply:
I don't think this new code is going to work. Also, if the thread is reading the correct time every time it wakes up there is no need to have any drift value.
The original sampling thread wakes up every 10 ms, but reads the correct time every 100 ms, updating the elapsed time. The pseudocode is something like
```
crtTime = 0;
while(sleep_interruptable(samplingPeriod) == 0)
   crtTime += samplingPeriod
   if (100 ms or more have passed since last sync time)
      crtTime = j9time_current_time_millis() - startTime
      update time when last sync happened
```

Reply:
Sampler thread sleep for  max(_minSamplingPeriod, jitconfig->samplingFrequency) so if we put any value on command line greater then 10msec it will sleep for that time.
Or depend the state of the VM, sampling frequency in idle state is 1000 msec.

Both the implementation are essentially doing the same thing.

Sample thread is updating the elapsed time every iteration with sampling period which might not be the right current time and then after 100 msec update the elapsed time with the correct time i.e. sync.

Statistics thread is updating the elapsed time every time with the correct time.
More over it take into the account of the extra time taken by the thread to execute due to system latency and adjust the sleep window to accommodate this latency. If every time the thread wake up at right time then there will be no drift and not adjustment.